cmd/bootstrap/cmd/final_list.go

```diff
@@ -0,0 +1,238 @@
+
+	// write node-config.json with the new list of nodes to be used for the `finalize` command
+	writeJSON(fmt.Sprintf(flagOutdir, "node-config.json"), model.ToPublicNodeInfoList(mixedNodeInfos))
+}
+
+func validateNodes(nodes []model.NodeInfo, stakingNodes []model.NodeInfo) {
+	// check node count
+	if len(nodes) != len(stakingNodes) {
+		log.Error().Int("nodes", len(nodes)).Int("staked nodes", len(stakingNodes)).
+			Msg("staked node count does not match flow and parnter node count")
+	}
+
+	// check staked and collected nodes and make sure node ID are not missing
+	validateNodeIDs(nodes, stakingNodes)
+
+	// create map
+	var nodesByID map[flow.Identifier]model.NodeInfo
+	for _, node := range nodes {
+		nodesByID[node.NodeID] = node
+	}
+
+	// check node type mismatch
+	for _, stakedNode := range stakingNodes {
+		matchingNode, ok := nodesByID[stakedNode.NodeID]
+		if !ok {
+			log.Warn().Str("staked node id", stakedNode.NodeID.String()).
+				Msg("no matching node found in non-contract nodes list")
+			continue
+		}
+
+		// check node type and error if mismatch
+		if matchingNode.Role != stakedNode.Role {
+			log.Error().Str("staked node", stakedNode.NodeID.String()).
+				Str("staked node-type", stakedNode.Role.String()).
+				Str("node", matchingNode.NodeID.String()).
+				Str("node-type", matchingNode.Role.String()).
+				Msg("node type does not match")
+		}
+
+		// No need to error adderss, and public keys as they may be different
+		// we only keep node-id constant through out sporks
+
+		// check address match
+		if matchingNode.Address != stakedNode.Address {
+			log.Warn().Str("staked node", stakedNode.NodeID.String()).
+				Str("node", matchingNode.NodeID.String()).
+				Msg("address do not match")
+		}
+
+		// flow nodes contain private key info
+		if matchingNode.NetworkPubKey().String() != "" {
+			// check networking pubkey match
+			matchNodeKey := matchingNode.NetworkPubKey().String()
+			stakedNodeKey := stakedNode.NetworkPubKey().String()
+
+			if matchNodeKey != stakedNodeKey {
+				log.Warn().Str("staked network key", stakedNodeKey).
+					Str("network key", matchNodeKey).
+					Msg("networking keys do not match")
+			}
+		}
+
+		// flow nodes contain priv atekey info
+		if matchingNode.StakingPubKey().String() != "" {
+			matchNodeKey := matchingNode.StakingPubKey().String()
+			stakedNodeKey := stakedNode.StakingPubKey().String()
+
+			if matchNodeKey != stakedNodeKey {
+				log.Warn().Str("staked staking key", stakedNodeKey).
+					Str("staking key", matchNodeKey).
+					Msg("staking keys do not match")
+			}
+		}
+	}
+}
+
+// validateNodeIDs will go through both sets of nodes and ensure that no node-id
+// are missing. It will log all missing node ID's and throw an error.
+func validateNodeIDs(collectedNodes []model.NodeInfo, stakedNodes []model.NodeInfo) {
+
+	// go through staking nodes
+	invalidStakingNodes := make([]model.NodeInfo, 0)
+	for _, node := range stakedNodes {
+		if node.NodeID.String() == "" {
```

Comment:
Sorry, maybe I was clear. The validation is not only to check whether NodeID is missing, but also see if there is collected node which is not a staked node, and if there is a staked node which is not a staked nodes. 

For instance say collectedNodes has nodes with ID `A, B, C`, and stakedNodes has `B, C, D`, and then we have mismatching collected nodes `A` and mismatching staked nodes `D`.

If there is either mismatching collected nodes or mismatching staked nodes, the program will print fatal error with the miss matching nodes.

Does it make sense?

Reply:
Ahh yes I see. Ok my bad,  I'll fix that now.